### PR TITLE
[New Feature] Send Interactive Reply Buttons Message

### LINKED ISSUE #8: Implement interactive message API.
The `messages_api` allows users to send interactive messages. 

We need to implement the functions defined in `messages.rb` 

https://github.com/ignacio-chiazzo/ruby_whatsapp_sdk/blob/5cf41178f74ffaf7f69a5832af4f5cc517466f17/lib/whatsapp/api/messages.rb#L181-L190

Read the [FB documentation here](https://developers.facebook.com/docs/whatsapp/cloud-api/reference/messages#contacts-object).

### PR DESCRIPTION
**Pull request by alienware**
# WHAT
* Adds classes for,
  * `Interactive`
  * `InteractiveHeader`
  * `InteractiveBody`
  * `InteractiveFooter`
  * `InteractiveAction`
  * `InteractiveActionButton`
* Adds example to send interactive reply buttons message

# WHY
Fixes partially #8

### CODE REVIEW

**Comment by alienware** [comment]
@ignacio-chiazzo Please check out and share your feedback regarding the implementation of interactive reply button messages. Let me know what needs to be done to make it ready.

File: lib/whatsapp_sdk/api/messages.rb [hunk 1]
```diff
@@ -363,9 +363,52 @@ def send_contacts(
...
+      # @param sender_id [Integer] Sender' phone number.
+      # @param recipient_number [Integer] Recipient' Phone number.
+      # @param interactive [Interactive] Interactive.
+      # @param interactive_json [Json] The interactive as a Json. If you pass interactive_json, you can't pass interactive.
+      # @param message_id [String] The id of the message to reply to.
+      # @return [WhatsappSdk::Api::Response] Response object.
+      sig do
+        params(
+          sender_id: Integer, recipient_number: Integer, interactive: T.nilable(T.untyped),
+          interactive_json: T.nilable(String), message_id: T.nilable(String),
```
Comment (ignacio-chiazzo) [inline lib/whatsapp_sdk/api/messages.rb:379]: Could `interactive_json` be a JSON type instead? 

`T::Hash[T.untyped, T.untyped]` ?

File: lib/whatsapp_sdk/api/messages.rb [hunk 2]
```diff
@@ -363,9 +363,52 @@ def send_contacts(
...
+        ).returns(WhatsappSdk::Api::Response)
+      end
+      def send_interactive_reply_buttons(
```
Comment (ignacio-chiazzo) [inline lib/whatsapp_sdk/api/messages.rb:382]: Let's add a test in the `messages_test.rb` file. You will need to mock the output of the API. See [Template test examples](https://github.com/ignacio-chiazzo/ruby_whatsapp_sdk/blob/db9a4c43df4627a8780daefcb85eecd888a60125/test/whatsapp/api/messages_test.rb#L501:L650).

File: lib/whatsapp_sdk/resource/interactive.rb [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+          ListMessage = new("list")
+          ReplyButton = new("button")
+          SingleProductMessage = new("product")
+          MultiProductMessage = new("product_list")
+        end
+      end
+
+      # Returns the Interactive type of message you want to send.
+      #
+      # @returns type [String]. Supported Options are list, button, product and product_list.
```
Comment (ignacio-chiazzo) [inline lib/whatsapp_sdk/resource/interactive.rb:70]:
```suggestion
      # @returns type [Type]. Supported Options are list, button, product and product_list.
```

File: lib/whatsapp_sdk/resource/interactive.rb [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        raise InvalidType, type
+      end
+
+      sig { void }
+      def validate_action
+        raise InvalidActionButtonsCount, action.buttons unless (1..3).include?(action.buttons.length)
+
+        action_button_ids = action.buttons.map(&:id)
+        raise InvalidActionButtonsId, action.buttons unless action_button_ids.length.eql?(action_button_ids.uniq.length)
+      end
```
Comment (ignacio-chiazzo) [inline lib/whatsapp_sdk/resource/interactive.rb:145]: Could you create a class `interactive_test.rb` and add all these validations?

File: lib/whatsapp_sdk/resource/interactive_action.rb [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+      private
+
+      sig { void }
+      def validate
+        validate_buttons
+      end
+
+      sig { void }
+      def validate_buttons
```
Comment (ignacio-chiazzo) [inline lib/whatsapp_sdk/resource/interactive_action.rb:65]: Please, let's add tests here.

File: lib/whatsapp_sdk/version.rb [hunk 6]
```diff
@@ -2,5 +2,5 @@
 # frozen_string_literal: true
 
 module WhatsappSdk
-  VERSION = "0.7.3"
+  VERSION = "0.7.4"
```
Comment (ignacio-chiazzo) [inline lib/whatsapp_sdk/version.rb:5]: Let's bump the version in a separate PR.

**Review by ignacio-chiazzo — COMMENTED** [review]
Hey! 

Thanks for implementing this. It's been a feature that I planned to add for a long time. Great work!

File: lib/whatsapp_sdk/api/messages.rb [hunk 7]
```diff
@@ -363,9 +363,52 @@ def send_contacts(
...
+      #
+      # @param sender_id [Integer] Sender' phone number.
+      # @param recipient_number [Integer] Recipient' Phone number.
+      # @param interactive [Interactive] Interactive.
+      # @param interactive_json [Json] The interactive as a Json. If you pass interactive_json, you can't pass interactive.
+      # @param message_id [String] The id of the message to reply to.
+      # @return [WhatsappSdk::Api::Response] Response object.
+      sig do
+        params(
+          sender_id: Integer, recipient_number: Integer, interactive: T.nilable(T.untyped),
```
[the remark this answers is not in the document]
Comment (alienware) [reply lib/whatsapp_sdk/api/messages.rb:378]: I will go ahead and fix the type of `interactive` parameter. I was following the convention from `send_template` to allow `interactive_json` which rejects `interactive` if both are present.

[the remark this answers is not in the document]
Comment (ignacio-chiazzo) [reply lib/whatsapp_sdk/api/messages.rb]: oh you're right, it should be Nilable and with the type Interactive

File: lib/whatsapp_sdk/resource/interactive_action_button.rb [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+        sig { params(type: String).void }
+        def initialize(type)
+          @message = T.let(
+            "invalid type #{type} for interactive. type should be reply only.",
+            String
+          )
+          super(message)
+        end
+      end
```
Comment (ignacio-chiazzo) [inline lib/whatsapp_sdk/resource/interactive_action_button.rb:23]: How do you feel if we extract this class into a generic error in its own file like this? 

```ruby
class ErrorWithMessage < StandardError
  extend T::Sig

  sig { returns(String) }
  attr_reader :message

  sig { params(message: String).void }
  def initialize(message)
    @message = T.let(
      "invalid type #{type} for interactive. type should be reply only.",
      String
    )
    super(message)
  end
end
```

Then in each file, we can reuse it.

File: lib/whatsapp_sdk/resource/interactive_action_button.rb [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+        raise InvalidTitle, title.length
+      end
+
+      sig { void }
+      def validate_id
+        return unless id
+
+        # TODO: Fix sorbet complaining NoMethodError for nil
+        id = id.dup
```
Comment (ignacio-chiazzo) [inline lib/whatsapp_sdk/resource/interactive_action_button.rb:117]: I'm surprised Sorbet doesn't know we check for nilability a line above. 

Could you use `T.must` instead?

File: lib/whatsapp_sdk/resource/interactive_action_button.rb (hunk 8 above)
Comment (ignacio-chiazzo) [reply lib/whatsapp_sdk/resource/interactive_action_button.rb:23]: I'm ok with doing it in a separate PR if it's getting too big.

File: lib/whatsapp_sdk/resource/interactive_action.rb [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+      # @returns buttons [Array<InteractiveActionButton>] .
+      sig do
+       returns(T::Array[InteractiveActionButton])
+      end
+      attr_accessor :buttons
+
+      # TODO: attr_accessor :button
+      # TODO: attr_accessor :sections
+      # TODO: attr_accessor :catalog_id
+      # TODO: attr_accessor :product_retailer_id
```
Comment (ignacio-chiazzo) [inline lib/whatsapp_sdk/resource/interactive_action.rb:20]: Are you planning to add them in a follow up PR?

File: lib/whatsapp_sdk/resource/interactive_action_button.rb (hunk 8 above)
Comment (alienware) [reply lib/whatsapp_sdk/resource/interactive_action_button.rb:23]: Help me understand the purpose of `ErrorWithMessage`, as I see it is simply `StandardError` as it doesn't have anything other than `message`. Should we DRY out specifically `InvalidType` error from multiple definitions?

File: lib/whatsapp_sdk/resource/interactive_action_button.rb (hunk 9 above)
Comment (alienware) [reply lib/whatsapp_sdk/resource/interactive_action_button.rb:117]: This is the my first project using Sorbet, let me read up more on how to crack it.

File: lib/whatsapp_sdk/resource/interactive_action.rb (hunk 10 above)
Comment (alienware) [reply lib/whatsapp_sdk/resource/interactive_action.rb:20]: These should definitely be done in a follow up PR as the scope of this PR is to add support for reply buttons only.

I see myself working of the usecases of catalog and products soon. When I do, I can lend a helping hand then to complete this module.

Comment (ignacio-chiazzo) [reply lib/whatsapp_sdk/resource/interactive_action.rb:20]: Thank you

File: lib/whatsapp_sdk/resource/interactive_action_button.rb (hunk 8 above)
Comment (ignacio-chiazzo) [reply lib/whatsapp_sdk/resource/interactive_action_button.rb:23]: Yeah, the idea is we can reuse it since it's basically the same across all classes but just a different message.

Comment (alienware) [reply lib/whatsapp_sdk/resource/interactive_action_button.rb:23]: Can we use something like
```ruby
module WhatsappSdk
  module Resource
    module Error
      class InvalidType < WhatsappSdk::Error
        extend T::Sig

        sig { returns(String) }
        attr_reader :message

        sig { params(type: String, klass: String, valid_types: Array).void }
        def initialize(type, klass, valid_types)
          @message = T.let(
            "invalid type #{type} for #{klass}. type should be #{valid_types.map(&:serialize).join(', ')} only.",
            String
          )
          super(message)
        end
      end
    end
  end
end
```

Although the more I'm trying it out, `InvalidType` seems to be an error we should never need as sorbet `T::Enum` should take care of the type and incorrect value. Should I go ahead and remove it?

Comment (alienware) [reply lib/whatsapp_sdk/resource/interactive_action_button.rb]: I've taken a stab at it, please check if the move was in the right direction.

File: lib/whatsapp_sdk/resource/interactive_action_button.rb (hunk 9 above)
Comment (alienware) [reply lib/whatsapp_sdk/resource/interactive_action_button.rb]: For some weird reason, `id` was coming as `nil`, even when `self.id` was set. I also realized while testing that `id` cannot be empty and so I've done away with the nilability.

**Review by ignacio-chiazzo — APPROVED** [review]
It looks good!

Let's add the TODO comments in a separate PR. After that, I will test and cut a new release.
